File: Qytetet/Qytetet_JAVA/src/interfazTextualQytetet/ControladorQytetet.java
package interfazTextualQytetet;
import modeloqytetet.*;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Comparator;
import java.util.Map;
public class ControladorQytetet {
	private Qytetet juego;
	private Jugador jugador;
	private Casilla casilla;
	private VistaTextualQytetet vista;
	
	public ControladorQytetet() {
		jugador = null;
		casilla = null;
		vista = new VistaTextualQytetet();
	};
	
	public void desarrolloJuego() {
		boolean puedeJugar = true, finalizado = false;
		vista.mostrar("------------------------------------");
		String rs = "Es el turno de " + jugador.getNombre() + ", actualmente se encuentra en la casilla " + casilla.getNumeroCasilla();
		if(casilla.getTipo() == TipoCasilla.CALLE) {
			rs += " con nombre " + casilla.getTitulo().getNombre();
		}else {
			rs += " de tipo " + casilla.getTipo();
		}
		vista.mostrar(rs);
		
		if(jugador.getEncarcelado()) {
			int valor = vista.menuSalirCarcel();
			juego.intentarSalirCarcel(valor == 0 ? MetodoSalirCarcel.PAGANDOLIBERTAD : MetodoSalirCarcel.TIRANDODADO);
			if(jugador.getEncarcelado()) {
				vista.mostrar("Desgraciadamente el jugador sigue en la carcel.");
				puedeJugar = false;
			}else{
				vista.mostrar("Afortunadamente consiguio salir sin muchos problemas.");
			}
		}
		
		if(puedeJugar){
			  boolean noTienePropietario = juego.jugar();
			  Casilla desplazamiento = jugador.getCasillaActual();
			  String ra = "Tras lanzar el dado, el jugador se desplazo hacia ";
			  if(desplazamiento.getTipo() == TipoCasilla.CALLE) {
				  ra += desplazamiento.getTitulo().getNombre();
			  }else {
				  ra += " la casilla de tipo " + desplazamiento.getTipo();
			  }
			  vista.mostrar(ra + " (" + desplazamiento.getNumeroCasilla() + ").");

			  if(jugador.estaBancarrota()) {
				  finalizado = true;
			  }else if(!jugador.getEncarcelado()) {
				  switch(desplazamiento.getTipo()) {
				  	case SORPRESA:
				  		vista.mostrar("Se trata de una carta sorpresa " + desplazamiento.getTipo() + " -> " + juego.getCartaActual().getDescripcion());
				  		noTienePropietario = juego.aplicarSorpresa();
				  	break;
				  	case CALLE:
				  		vista.mostrar(desplazamiento.toString());
			  		break;
				  	default:break;
				  }
				  if(!jugador.getEncarcelado()) {
					  if(!noTienePropietario){
						  boolean comprar = vista.elegirQuieroComprar();
						  if(comprar) {
							  boolean compra = juego.comprarTituloPropiedad();
							  if(compra) {
								  vista.mostrar("El usuario compró " + desplazamiento.getTitulo().getNombre());	  
							  }else {
								  vista.mostrar("No se pudo realizar la compra.");
							  }
						  }
					  }
					  vista.mostrar(jugador.toString());
	
					  if(jugador.estaBancarrota()) {
						  finalizado = true;
					  }else if(jugador.tengoPropiedades()) {
						  int opcion = vista.menuGestionInmobiliaria();
						  while(opcion != 0) {
							  ArrayList<Casilla> casillas = new ArrayList<Casilla>();
							  for(TituloPropiedad titulo : jugador.propiedades) {
								  casillas.add(titulo.casilla);
							  }
							  casilla = this.elegirPropiedad(casillas);
							  boolean posible = true;
							  switch(opcion) {
							  	case 1: posible = juego.edificarCasa(casilla); break;
							  	case 2: posible = juego.edificarHotel(casilla); break;
							  	case 3: posible = juego.venderPropiedad(casilla); break;
							  	case 4: posible = juego.hipotecarPropiedad(casilla); break;
							  	case 5: posible = juego.cancelarHipoteca(casilla); break;
							  	default:break;
							  }
							  if(!posible) {
								  vista.mostrar("No se pudo realizar la operacion");
							  }
							  opcion = vista.menuGestionInmobiliaria();
						  }
					  }
				  }else {
					  vista.mostrar(jugador.toString());
				  }
			  }
		}
		
		if(!finalizado) {
			juego.siguienteJugador();
			jugador = juego.getJugadorActual();
			casilla = jugador.getCasillaActual();
			if(jugador.estaBancarrota()) {
				finalizado = true;
			}else {
				this.desarrolloJuego();
			}
		}
		
		if(finalizado){
			vista.mostrar("El juego ha acabado, un jugador esta en bancarrota.");
			ArrayList<Map.Entry<String, Integer>> ranking = juego.obtenerRanking();
			for(Map.Entry<String, Integer> person : ranking) {
				vista.mostrar(person.getKey() + " con un capital " + person.getValue());
			}
		}
	};
	
	private Casilla elegirPropiedad(ArrayList<Casilla> propiedades) {
	  ArrayList<String> nombres = new ArrayList<String>();
	  for(Casilla casilla : propiedades) {
		  nombres.add(casilla.getTitulo().getNombre());
	  }
	  int opcion = vista.menuElegirPropiedad(nombres);
	  return propiedades.get(opcion);
	};
	
	public void inicializacionJuego() {
		juego = Qytetet.getInstance();
		ArrayList<String> nombres = vista.obtenerNombreJugadores();
		juego.inicializarJuego(nombres);
		jugador = juego.getJugadorActual();
		casilla = jugador.getCasillaActual();
		vista.mostrar(juego.toString());
		this.desarrolloJuego();
	};
}

File: PDOO/Qytetet_RUBY/src/interfaz_textual_qytetet/ControladorQytetet.rb
#encoding: utf-8
require_relative '../modelo_qytetet/Qytetet'
require_relative 'VistaTextualQytetet'
module InterfazTextualQytetet
  class ControladorQytetet
    def initialize
      @juego = nil
      @jugador = nil
      @casilla = nil
      @vista = VistaTextualQytetet.new
    end
    
    def desarrollo_juego
      puede_jugar = true;
      finalizado = false;
      @vista.mostrar("------------------------------------")
      rs = "Es el turno de #{@jugador.nombre}, actualmente se encuentra en la casilla #{@casilla.numero_casilla}"
      if @casilla.tipo == ModeloQytetet::TipoCasilla::CALLE
        rs += " con nombre #{@casilla.titulo.nombre}"
      else
        rs += " de tipo #{@casilla.tipo}"
      end
      @vista.mostrar(rs)
      
      if @jugador.encarcelado
        valor = @vista.menu_salir_carcel
        metodo = nil
        if valor == 0
          metodo = ModeloQytetet::MetodoSalirCarcel::TIRANDODADO
        else
          metodo = ModeloQytetet::MetodoSalirCarcel::PAGANDOLIBERTAD
        end
        @juego.intentar_salir_carcel(metodo)
        if @jugador.encarcelado
          @vista.mostrar("Desgraciadamente el jugador sigue en la carcel.")
          puede_jugar = false
        else
          @vista.mostrar("Afortunadamente consiguio salir sin muchos problemas.");
        end
      end
      
      if puede_jugar
        no_tiene_propietario = @juego.jugar
        desplazamiento = @jugador.casilla_actual
        ra = "Tras lanzar el dado, el jugador se desplazo hacia "
        if desplazamiento.tipo == ModeloQytetet::TipoCasilla::CALLE
          ra += desplazamiento.titulo.nombre
        else
          ra += " la casilla de tipo #{desplazamiento.tipo}"
        end
        @vista.mostrar(ra)
        if !@jugador.encarcelado
          if !no_tiene_propietario
            comprar = @vista.elegir_quiero_comprar
            if comprar
              compra = @juego.comprar_titulo_propiedad
              if compra
                @vista.mostrar("El usuario compró #{desplazamiento.titulo.nombre}");
              else 
                @vista.mostrar("No se pudo realizar la compra.");
              end       
            end
            @vista.mostrar(@jugador.to_s)
            if @jugador.esta_bancarrota
              finalizado = true
            elsif @jugador.tengo_propiedades
              opcion = @vista.menu_gestion_inmobiliaria
              while opcion != 0 do
                posible = true
                case opcion
                  when 1
                    posible = @juego.edificar_casa(casilla)
                  when 2
                    posible = @juego.edificar_hotel(casilla)
                  when 3
                    posible = @juego.vender_propiedad(casilla)
                  when 4
                    posible = @juego.hipotecar_propiedad(casilla)
                  when 5
                    posible = @juego.cancelar_hipoteca(casilla)
                end
                if !posible
                  @vista.mostrar("No se pudo realizar la operacion");
                end
                opcion = @vista.menu_gestion_inmobiliaria
              end
            end
          end
        else
          @vista.mostrar(@jugador.to_s)
        end
      end
      
      if !finalizado 
        @juego.siguiente_jugador
        @jugador = @juego.jugador_actual
        @casilla = @jugador.casilla_actual
        if @jugador.esta_bancarrota
          finalizado = true
        else
          desarrollo_juego
        end
      end
      
      if finalizado
        @vista.mostrar("El juego ha acabado, un jugador esta en bancarrota.")
        ranking = @juego.obtener_ranking
        ranking.each{ |p|
          @vista.mostrar("#{p[0]} con un capital de #{p[1]}")
        }
      end
    end
    
    def elegir_propiedad(propiedades)
      nombres = Array.new
      propiedades.each{ |p| nombres << p.nombre }
      opcion = @vista.menu_elegir_propiedad(nombre)
      return propiedades[opcion]
    end
    
    def inicializacion_juego
      @juego = ModeloQytetet::Qytetet.instance
      nombres = @vista.obtener_nombre_jugadores
      @juego.inicializar_juego(nombres)
      @jugador = @juego.jugador_actual
      @casilla = @jugador.casilla_actual
      @vista.mostrar(@juego.to_s)
      desarrollo_juego
    end
    
    private :elegir_propiedad
  end
end